weighted_levenshtein/clev.pyx

```diff
@@ -132,12 +132,29 @@ cdef inline DTYPE_t row_insert_range_cost(
 
 # End Array2D
 
+def convert_string_to_int_array(char* str):
+    s = str.encode('utf-8').decode('utf-8')
+    intarr = np.empty([len(s), ], dtype=int)
+    for i, c in enumerate(s):
+        intarr[i] = ord(c)
+    return intarr
 
 cdef inline unsigned char str_1_get(unsigned char* s, Py_ssize_t i) nogil:
     """
     Takes an index of a 1-indexed string
     and returns that character
     """
+    if i - 1 < 0:
+        return 0
+    return s[i - 1]
+
+cdef inline int int_array_1_get(int[:] s, Py_ssize_t i) nogil:
+    """
+    Takes an index of a 1-indexed int array
+    and returns that number
+    """
+    if i - 1 < 0:
```

Comment:
I think if i < 1, this is indicative of a bug and we would want this to crash instead of silently returning 0

Reply:
It's there, because this was the fix for Windows crashing. 

On line 279 you have this:
`transpose_costs[int_array_1_get(str1, k), int_array_1_get(str1, i)]`
But the value of `k` comes from line 265:
`k = da[char_j]`
Since `da` is filled with zeroes initially, that means that k will always be 0 in the first iteration, which means that you'll just get some random memory. And then you put that random memory into `transpose_costs`.
This caused a crash on Windows, but I'm not sure how did it work on Linux.

Am I correct here? Is `k` intended to be zero? If so, what did it do on Linux?

Reply:
Thank you for the explanation, I believe you are correct, the bug is with `k` being 0 initially,

It has been a few years since I looked at this code, so my understanding might be fuzzy. Please take everything with a grain of salt.

I believe what `da` stores is the rightmost index where a character was seen. This is used to find the rightmost swap location for `char_j`, which is `da[char_j]`. 

`da` is initialized with all 0s, because 0 means that the character has not been seen so far.
Thus when we try to do `k = da[char_j]`, since it hasn't seen `char_j` yet, it returns a 0. Then, when we try to do `str_1_get(str1, k)`, it segfaults, since the smallest index we can get is 1.

0 is a sentinel value, but the code right now treats it like a regular index.

I believe the correct solution for this would be to rewrite this section:
```
            Array2D_n1_at(d, i, j)[0] = min(
                Array2D_n1_get(d, i - 1, j - 1) + cost,                  # equal/substitute
                Array2D_n1_get(d, i, j - 1) + insert_costs[char_j],    # insert
                Array2D_n1_get(d, i - 1, j) + delete_costs[char_i],    # delete
                Array2D_n1_get(d, k - 1, l - 1) +                        # transpose
                    col_delete_range_cost(d, k + 1, i - 1) +                      # delete chars in between
                    transpose_costs[str_1_get(str1, k), str_1_get(str1, i)] +   # transpose chars
                    row_insert_range_cost(d, l + 1, j - 1)                        # insert chars in between
            )
```

as 

```
            substitute_cost = Array2D_n1_get(d, i - 1, j - 1) + cost
            insert_cost = Array2D_n1_get(d, i, j - 1) + insert_costs[char_j]
            delete_cost = Array2D_n1_get(d, i - 1, j) + delete_costs[char_i
            if k <= 0: 
                # char_j hasn't been seen yet, so nothing to swap
                transpose_cost = DTYPE_MAX
            else: 
                # char_j has been seen, swap with char_i
                transpose_cost = 
                    Array2D_n1_get(d, k - 1, l - 1) +                        # transpose
                        col_delete_range_cost(d, k + 1, i - 1) +                      # delete chars in between
                        transpose_costs[char_j, char_i] +   # transpose chars
                        row_insert_range_cost(d, l + 1, j - 1)                        # insert chars in between
            
            Array2D_n1_at(d, i, j)[0] = min(
                substitute_cost,
                insert_cost,
                delete_cost,
                transpose_cost
            )
```

Note that I changed `transpose_costs[str_1_get(str1, k), str_1_get(str1, i)]` to `transpose_costs[char_j, char_i]` because 
- `str_1_get(str1, i)` is just `char_i`
- `str_1_get(str1, k)` is just `char_j`, because in this branch `k > 0`, which means that `k` is the rightmost index where `char_j` can be found. Thus getting index `k` of the string is just `char_j`.

I'm somewhat confident this is correct, but could you please run the unit tests to make sure? Thanks!

If the unit tests pass with the changes above, then I think the `if i - 1 < 0: return 0` in `int_array_1_get` will no longer be necessary.

Reply:
Yep, it did work!

Reply:
Awesome, thanks for verifying!